DOC Add common pitfalls: data leakage (#18426)
scikit-learn/scikit-learn: doc/pitfalls/data_leakage.rst

.. _data_leakage:

============
Data leakage
============

Data leakage occurs when information that would not be available at prediction

> time, is used when building the model. This results in overly optimstic

Reviewer: ```suggestion
time, is used when building the model. This results in overly optimistic
```

performance estimates, for example from :ref:`cross validation

> <cross_validation>`, but poorer performance when the model is asked to predict

Reviewer: Maybe
```suggestion
<cross_validation>`, but poorer performance when the model is used
```
since it's not sentient yet ) Though maybe I'm misinterpreting the wording.

    >>> from sklearn.model_selection import cross_val_score
    >>> X_impute = SimpleImputer().fit_transform(X_missing)
    >>> scores = cross_val_score(GradientBoostingRegressor(random_state=1),
    ...                          X_impute, y)
    >>> print(f"Mean accuracy: {scores.mean():.3f}+/-{scores.std():.2f}")
    Mean accuracy: 0.412+/-0.06

**Right**

Using a :class:`~sklearn.pipeline.Pipeline` to chain together the imputation
and model ensures that only the train data subset is using for imputation.
This results in a much lower accuracy::

    >>> from sklearn.pipeline import make_pipeline
    >>> pipeline = make_pipeline(SimpleImputer(),
    ...                          GradientBoostingRegressor(random_state=1))
    >>> scores = cross_val_score(pipeline, X_missing, y)
    >>> print(f"Mean accuracy: {scores.mean():.3f}+/-{scores.std():.2f}")
    Mean accuracy: 0.421+/-0.07

Use pipelines
=============

You may have noticed a common theme in our examples. Both the 'Right' examples
use the :ref:`pipeline <pipeline>`, which helps prevent data leakage by
only using the training data to calculate preprocessing statistics. Conversely,

> both the 'Wrong' examples used the :term:`fit_transform` method.

Reviewer: ```suggestion
both the 'Wrong' examples used the :term:`fit_transform` method of
individual estimators.
```

.. _data_leakage:

============
Data leakage
============

Data leakage occurs when information that would not be available at prediction
time, is used when building the model. This results in overly optimsitic

> performance estimates, for example from :ref:`cross validation

Reviewer 2: ```suggestion
performance estimates, for example from :ref:`cross-validation
```

<cross_validation>`, but poorer performance when the model is used
on actually novel data, for example during production.

> A common cause is not keeping the test and train data subsets separate. Test

Reviewer 3: I would write something like:

> The general rule is to never call `fit` on some testing data. While this may sound obvious, this is easy to miss in some cases, for example when applying some pre-processing steps.

data should never be used to make choices about the model but this may
accidently occur during :ref:`preprocessing` or :ref:`grid_search`.

Although both train and test data subsets should undergo the same preprocessing
transformation, it is important that stateful transformations only use the
training subset to determine the 'state'. For example, if you have a
normalization step whereby you divide by the average value, the average should
be the average of the train subset, **not** the average of all the data. If the
test subset was included in the average calculation, information from the test
subset is influencing the model. Other types of preprocessing such as
:ref:`impute` and :ref:`polynomial_features` should also only utilize train

> data.

Reviewer 2: I would add a note that this anti-pattern is to call fit_transform(X_test) and not transform(X_test).

Reviewer 2: or link to specific examples just below so a clear example is a click away

Reviewer 3: I would go further and not that the anti-pattern is to call `fit(X_test)` and thus this applies to `fit_transform()` as well ;)

Including the test data when finding the best model hyperparameters will
also inadvertantly introduce information from the test data into the model.
Practically, in scikit-learn, this means that only the train data subset

> should be fed into the `fit` method of :ref:`hyper_parameter_optimizers`.

Author: yes true, I can amend to say that when fitting the model or tuning hyperparameters, never feed test data into `fit`

Some examples of common data leakage pitfalls are detailed below.

Data leakage during feature selection
=====================================

A number of :ref:`feature_selection` functions are available in scikit-learn.
They can help remove irrelevant, redundant and noisy features as well as
improve your model build time and performance. Feature selection should
**only** use the training data. Including the test data in feature selection
will optimistically bias your model.

> To demonstrate we will create a binary classification problem with

Reviewer 2: ```suggestion
To demonstrate this we will create a binary classification problem with
```

:func:`~sklearn.model_selection.cross_val_score` ensures that only the
training data is used for feature selection (and when fitting our model).
The test data will only be used for calculating the accuracy score. The
score is now what we would expect for the data, close to chance::

    >>> from sklearn.pipeline import make_pipeline
    >>> pipeline = make_pipeline(SelectKBest(k=25),
    ...                          GradientBoostingClassifier(random_state=1))
    >>> scores = cross_val_score(pipeline, X, y)
    >>> print(f"Mean Accuracy: {scores.mean()}")
    Mean Accuracy: 0.495

Data leakage during imputation
==============================

There are a number of methods to impute missing values in data. For example,
:class:`~sklearn.impute.SimpleImputer` allows you to replace missing values
with the mean of that feature. Only the train data should be used to
calculate this mean value as including the test data in the mean calculation
will introduce information about the test data into the model.

To demonstrate this, we will use the :ref:`diabetes_dataset` and
artificially introduce missing values. We introduce a large number of

> missing values to make the effect more pronounced::

Reviewer 3: Maybe a more realistic way that would also make the effect more pronounced would be to introduce not-missing-at-random values (i.e. missing values that depend on X and y)

Reviewer 3: (I haven't tried, maybe this isn't a great idea.)

If this were a classification dataset, one way to do it would be to introduce missing values (in a lot of features) for a bunch of samples from the same class. For a regression problem you can do that for target values within a given range.

Reviewer 3: I gave it a quick try and it doesn't seem to be super compelling. I think it's because the the SimpleImputer treats missing values as "missing at random" anyway. It would probably have some effect one the HistGradientBoosting which treat nans as non-random, but to illustrate it we would need to pass X_test to `fit` which would make no sense.

Author: I tried with several datasets and also creating a new dataset and I could not find a compelling example. And I also got mixed up on this one as the performance when only using the train data is better! Suggestions welcome...

    >>> import numpy as np
    >>> from sklearn.datasets import load_diabetes
    >>> X, y = load_diabetes(return_X_y=True)

>     >>> rng = np.random.RandomState(42)

Reviewer 2: don't use want to use

from numpy.random import default_rng

as above?

Reviewer 3: I'd rather not, we have zero use of `default_rng` throughout the codebase.

Also wouldn't this make results non-reproducible (and CI would complain)?

Reviewer 3: BTW this fails anyway: default_rng returns a Generator, and we don't support them.

```py
In [1]: from numpy.random import default_rng                                                                           

In [2]: from sklearn.datasets import make_classification                                                               
make_
In [3]: make_classification(random_state=default_rng)

ValueError: Generator(PCG64) at 0x7F4B13886C70 cannot be used to seed a numpy.random.RandomState instance
```

Author: Changed to use `np.random.RandomState` as in other examples.

Reviewer 4: another reason is that `default_rng` requires numpy 1.17 or above, scikit-learn only requires numpy 1.13

>     >>> n_samples = X.shape[0]
>     >>> n_features = X.shape[1]

Reviewer 2: ```suggestion
    >>> n_samples, n_features = X.shape
```

    >>> n_missing = int(n_samples * 0.75)
    >>> missing_samples = np.zeros(n_samples, dtype=np.bool)

>     >>> missing_samples[: n_missing] = True

Reviewer 2: ```suggestion
    >>> missing_samples[:n_missing] = True
```

    >>> rng.shuffle(missing_samples)
    >>> missing_features = rng.randint(0, n_features, n_missing)
    >>> X_missing = X.copy()
    >>> X_missing[missing_samples, missing_features] = np.nan

**Wrong**

Using all the data to calculate the feature means, to replace the missing

> values with, results in a very high accuracy::

Reviewer 2: ```suggestion
values with, results in a very high R2 score (chance level is 0)::
```

    >>> from sklearn.impute import SimpleImputer
    >>> from sklearn.ensemble import GradientBoostingRegressor
    >>> from sklearn.model_selection import cross_val_score
    >>> X_impute = SimpleImputer().fit_transform(X_missing)
    >>> scores = cross_val_score(GradientBoostingRegressor(random_state=1),
    ...                          X_impute, y)
    >>> print(f"Mean R^2: {scores.mean():.3f}+/-{scores.std():.2f}")
    Mean R^2: 0.412+/-0.06

**Right**

Using a :class:`~sklearn.pipeline.Pipeline` to chain together the imputation
and model ensures that only the train data subset is using for imputation.

> This results in a much lower accuracy::

Reviewer 2: ```suggestion
This results in a much lower R2 score::
```

    >>> from sklearn.pipeline import make_pipeline
    >>> pipeline = make_pipeline(SimpleImputer(),
    ...                          GradientBoostingRegressor(random_state=1))
    >>> scores = cross_val_score(pipeline, X_missing, y)
    >>> print(f"Mean R^2: {scores.mean():.3f}+/-{scores.std():.2f}")

>     Mean R^2: 0.421+/-0.07

Reviewer 2: 0.421 vs 0.412 above?

Author: whoops!

Use pipelines
=============

You may have noticed a common theme in our examples. Both the 'Right' examples

> use the :ref:`pipeline <pipeline>`, which helps prevent data leakage by

Reviewer 3: ```suggestion
use a :ref:`pipeline <pipeline>`, which helps prevent data leakage by
```

only using the training data to calculate preprocessing statistics. Conversely,
both the 'Wrong' examples used the :term:`fit_transform` method of the

> preprocessing estimator.

Reviewer 3: It's not really about `fit_transform` but mainly about using `fit` on the test data.
`fit_transform` is a subcase of that.

Care needs to be taken when using the :term:`fit_transform` method. This is
because it combines the `fit` method, which should
only be performed on the train subset, and the :term:`transform` method which
is generally performed on the whole dataset, as the train and test subsets
should be preprocessed in the same way. Scikit-learn pipelines ensure that

> the appropriate method is performed on the correct data subset.

Reviewer 2: You can basically saying that fit_transform(X) is bad
but I would also highlight that fit_transform(X_test) is also wrong 
as it can harm performance

time is used when building the model. This results in overly optimsitic
performance estimates, for example from :ref:`cross validation
<cross_validation>`, and thus poorer performance when the model is used
on actually novel data, for example during production.

A common cause is not keeping the test and train data subsets separate.
Test data should never be used to make choices about the model.
The general rule is to never call `fit` on testing data. While this may
sound obvious, this is easy to miss in some cases, for example when applying
some pre-processing steps.

Although both train and test data subsets should undergo the same preprocessing
transformation, it is important that these transformations are only learnt
from the training data. For example, if you have a
normalization step where you divide by the average value, the average should
be the average of the train subset, **not** the average of all the data. If the
test subset was included in the average calculation, information from the test
subset is influencing the model. Other types of preprocessing such as
:ref:`impute` (see :ref:`data_leakage_imputation` below) and

> :ref:`polynomial_features` should also only utilize train data.

Reviewer 5: PolynomialFeatures only depends on the shape of the training data, AFAIK, not its values, so it doesn't apply here.

Processing such as decomposition and kernel approximation can be affected. The most problematic transformers for leakage, however, are those that are supervised.
